src/flow/styled.js

```diff
@@ -0,0 +1,122 @@
+// @flow
+import type {
+  UnconditionalStyles,
+  ConditionalStylesSpec,
+  LiteralStylesSpec,
+  WhitelistedPropNames,
+  ZacsViewFunction,
+  ZacsTextFunction,
+} from './components'
+
+export type Component<T> = React$ComponentType<T>
+
+// NOTE: $Partial is ncessary to make the new props optional. $Shape doesn't do what you think it does
+// and using { [string]: style } doesn't work either
+type GetConditionalProps<ConditionalStyles> = $Partial<
+  $ObjMapConst<
+    ConditionalStyles,
+    mixed /* we're using mixed here because any truthy value is acceptable here */,
+  >,
+>
+// TODO: We could get fancy here and get the correct expected type based on style attribute
+type GetLiteralProps<LiteralStyles> = $Partial<$ObjMapConst<LiteralStyles, number | string>>
+
+// NOTE: It's probably possible to use some Flow magic to convert missing or null props to `{}` so that
+// we don't have to create a union function type in such an awful way
+type StyledFnWithConditionalLiteralStyles = <
+  OriginalProps,
+  ConditionalStyles: ConditionalStylesSpec,
+  LiteralStyles: LiteralStylesSpec,
+>(
+  component: Component<OriginalProps>,
+  unconditionalStyles: ?UnconditionalStyles,
+  conditionalStyles: ConditionalStyles,
+  literalStyles: LiteralStyles,
+  whitelistedProps: ?WhitelistedPropNames,
+) => Component<
+  $Exact<{
+    ...GetConditionalProps<ConditionalStyles>,
+    ...$Exact<OriginalProps>,
+    ...GetLiteralProps<LiteralStyles>,
+  }>,
+>
+
+type StyledFnWithConditionalStyles = <
+  OriginalProps,
+  ConditionalStyles: ConditionalStylesSpec,
+  LiteralStyles: LiteralStylesSpec,
+>(
+  component: Component<OriginalProps>,
+  unconditionalStyles: ?UnconditionalStyles,
+  conditionalStyles: ConditionalStyles,
+  literalStyles: ?null,
+  whitelistedProps: ?WhitelistedPropNames,
+) => Component<
+  $Exact<{
+    ...GetConditionalProps<ConditionalStyles>,
+    ...$Exact<OriginalProps>,
```

Comment:
`OriginalProps` may overwrite the conditional ones - is this how the implementation works?

Reply:
I don't think so, but this is the only way I could get typing to work :(